### PR TITLE
Feature/project control

### PR DESCRIPTION
**Pull request by Ayoush**
Changes:-

S10U4T1 | BE: Create API to list all projects
S10U6T1 | BE: Create DB to hold multiple project leads
S10U6T2 | BE: Create DB to hold multiple project users
S10U6T3 | BE: API to add a new project
S10U6T4 | BE: API to list all users while showing leads (should show only with admin and user role)
S10U6T5 | BE: API to list all users while adding users (should show all the users)
S10U7T1 | BE: Add model functions to update projects
S10U7T2 | BE: Add API to edit project details

### CODE REVIEW

File: apps/acqdat_api/lib/acqdat_api/entity-management/project.ex [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  defdelegate delete(project), to: ProjectModel
+
+  def update(project, params) do
+    project = project |> Repo.preload([:leads, :users])
+    params = update_version(params, project)
+    ProjectModel.update(project, params)
+  end
+
+  def create(attrs) do
+    verify_project(ProjectModel.create(project_create_attrs(attrs)))
```
Comment (arjun289) [inline apps/acqdat_api/lib/acqdat_api/entity-management/project.ex:16]: please use a pipe here

File: apps/acqdat_api/lib/acqdat_api/entity-management/project.ex [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  defp project_create_attrs(
+         %{
+           lead_ids: lead_ids,
+           user_ids: user_ids
+         } = params
+       ) do
+    params = params_extraction(params)
+
+    lead_ids = [0 | lead_ids]
+    user_ids = [0 | user_ids]
```
Comment (arjun289) [inline apps/acqdat_api/lib/acqdat_api/entity-management/project.ex:28]: Why this, I mean append 0?

File: apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  alias AcqdatApi.ImageDeletion
+  import AcqdatApiWeb.Helpers
+  import AcqdatApiWeb.Validators.EntityManagement.Project
+
+  plug AcqdatApiWeb.Plug.LoadOrg
+  plug AcqdatApiWeb.Plug.LoadProject when action in [:update, :delete, :show]
+
+  @doc """
+  This piece of code will be useful when we will implement Project role based listing
+  #case ProjectModel.check_adminship(Guardian.Plug.current_resource(conn)) do
```
Comment (arjun289) [inline apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex:14]: Please see how to add an example inside the doc, also separate the first line

File: apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        params =
+          case is_nil(params["image"]) do
+            true ->
+              params
+
+            false ->
+              ImageDeletion.delete_operation(project, "project")
+              add_image_url(conn, params)
+          end
```
Comment (arjun289) [inline apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex:54]: handle this in a separate function

File: apps/acqdat_api/lib/acqdat_api/entity-management/project.ex (hunk 2 above)
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/entity-management/project.ex:28]: It's a failsafe method while using integers in list since elixir interpret list with charlist and add random characters so to append zero ensure that correct list is being interpreted and 0 won't bother since ID can't be 0.

File: apps/acqdat_api/lib/acqdat_api/entity-management/project.ex (hunk 1 above)
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api/entity-management/project.ex:16]: Done

File: apps/acqdat_api/lib/acqdat_api_web/router.ex [hunk 5]
```diff
@@ -72,6 +74,9 @@ defmodule AcqdatApiWeb.Router do
     post("/projects/:project_id/entities", EntityManagement.EntityController, :update_hierarchy)
     get("/projects/:project_id/entities", EntityManagement.EntityController, :fetch_hierarchy)
 
+    resources "/projects", EntityManagement.ProjectController,
+      only: [:index, :create, :update, :delete, :show]
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api_web/router.ex:78]: @Ayoush  Here you need not have to specify all the actions inside only.
`resources "/projects", EntityManagement.ProjectController`
Just doing this, will be good

File: apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex (hunk 3 above)
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex:14]: Done

File: apps/acqdat_api/lib/acqdat_api_web/views/entity-management/asset_type_view.ex [hunk 6]
```diff
@@ -17,6 +16,7 @@ defmodule AcqdatApiWeb.EntityManagement.AssetTypeView do
       sensor_type_present: asset_type.sensor_type_present,
       sensor_type_uuid: asset_type.sensor_type_uuid,
       parameters: render_many(asset_type.parameters, AssetTypeView, "data_tree.json")
+      # org: render_one(asset_type.org, OrganisationView, "org.json")
```
Comment (BandanaPandey) [inline apps/acqdat_api/lib/acqdat_api_web/views/entity-management/asset_type_view.ex:19]: It this required @Ayoush ?
org as comment.
If not please remove it.

File: apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex (hunk 4 above)
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/controllers/entity-management/project_controller.ex:54]: Done

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/asset_type.ex [hunk 7]
```diff
@@ -111,7 +111,7 @@ defmodule AcqdatCore.Schema.EntityManagement.AssetType do
     |> put_change(:uuid, UUID.uuid1(:hex))
   end
 
-  defp add_slug(%Ecto.Changeset{valid?: true} = changeset) do
```
Comment (BandanaPandey) [inline apps/acqdat_core/lib/acqdat_core/entity-management/schema/asset_type.ex:114]: why did you remove valid changeset pattern match ?
It is intensional ?
@Ayoush

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex [hunk 8]
```diff
@@ -70,17 +85,46 @@ defmodule AcqdatCore.Schema.EntityManagement.Project do
     )
   end
 
-  defp add_uuid(%Ecto.Changeset{valid?: true} = changeset) do
```
Comment (BandanaPandey) [inline apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex:73]: Please don't remove validity pattern match @Ayoush

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex [hunk 9]
```diff
@@ -70,17 +85,46 @@ defmodule AcqdatCore.Schema.EntityManagement.Project do
...
+  defp add_uuid(changeset) do
     changeset
     |> put_change(:uuid, UUID.uuid1(:hex))
   end
 
-  defp add_slug(%Ecto.Changeset{valid?: true} = changeset) do
+  defp add_slug(changeset) do
```
Comment (BandanaPandey) [inline apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex:93]: same as above

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/sensor.ex [hunk 10]
```diff
@@ -9,6 +9,7 @@ defmodule AcqdatCore.Schema.EntityManagement.Sensor do
 
   use AcqdatCore.Schema
   alias AcqdatCore.Schema.EntityManagement.{Gateway, Organisation, Project, SensorsData}
+
```
Comment (BandanaPandey) [inline apps/acqdat_core/lib/acqdat_core/entity-management/schema/sensor.ex:12]: remove this line spacing

File: apps/acqdat_api/lib/acqdat_api_web/router.ex (hunk 5 above)
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/router.ex:78]: resources "/projects", EntityManagement.ProjectController this thing creates some extra unwanted routes like edit new which is not desired.

File: apps/acqdat_api/lib/acqdat_api_web/views/entity-management/asset_type_view.ex (hunk 6 above)
Comment (Ayoush) [reply apps/acqdat_api/lib/acqdat_api_web/views/entity-management/asset_type_view.ex:19]: Done

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/asset_type.ex (hunk 7 above)
Comment (Ayoush) [reply apps/acqdat_core/lib/acqdat_core/entity-management/schema/asset_type.ex:114]: Yes since these functions are also being used to add uuid and slug for embedded field and we are using those functions before validation hence validation for those comes nil hence that's why I removed.

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex (hunk 8 above)
Comment (Ayoush) [reply apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex:73]: Same reason as above we are already catching the error couple of places.

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex (hunk 9 above)
Comment (Ayoush) [reply apps/acqdat_core/lib/acqdat_core/entity-management/schema/project.ex:93]: same as above

File: apps/acqdat_core/lib/acqdat_core/entity-management/schema/sensor.ex (hunk 10 above)
Comment (Ayoush) [reply apps/acqdat_core/lib/acqdat_core/entity-management/schema/sensor.ex:12]: Done

File: apps/acqdat_api/lib/acqdat_api_web/router.ex (hunk 5 above)
Comment (BandanaPandey) [reply apps/acqdat_api/lib/acqdat_api_web/router.ex:78]: ohh, so you can also use this
`resources "/projects", EntityManagement.ProjectController, except: [:new, :edit]`
Thoughts ?
